src/Workspaces/Core/Desktop/Workspace/SQLite/SQLitePersistentStorage_WriteBatching.cs

```diff
@@ -69,100 +68,72 @@ internal partial class SQLitePersistentStorage
-            countdown.Wait();
+                    // and clear them from the queues so we don't process things multiple times.
+                    keyToWriteActions.Remove(key);
 
-            // If we're the thread that finally got the countdown to zero, then dispose of this
-            // count down and remove it from the dictionary (if it hasn't already been replaced
-            // by the next request).
-            if (lastSignal)
-            {
-                Debug.Assert(countdown.CurrentCount == 0);
+                    // Find the existing task responsible for writing to this queue.
+                    var existingWriteTask = keyToWriteTask.TryGetValue(key, out var task)
+                        ? task
+                        : SpecializedTasks.EmptyTask;
 
-                // Safe to call outside of lock.  Countdown is only given out to a set of threads
-                // that have incremented it.  And we can only get here once all the threads have
-                // been allowed to get past the 'Wait' point.  Only one of those threads will
-                // have lastSignal set to true, so we'll only dispose this once.
-                countdown.Dispose();
+                    if (writesToProcess.Count == 0)
+                    {
+                        // We have no writes of our own.  But there may be an existing task that
+                        // is writing out this queue.   Return this so our caller can wait for
+                        // all existing writes to complete.
+                        return existingWriteTask;
+                    }
 
-                using (await _writeQueueGate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
-                {
-                    // Remove the countdown if it's still in the dictionary.  It may not be if
-                    // another thread came in after this batch of threads completed, and it 
-                    // removed the completed countdown already.
-                    RemoveCountdownIfComplete(keyToCountdown, key);
+                    // We have our own writes to process.  Enqueue the task to write 
+                    // these out after the existing write-task for this queue completes.
+                    // 
+                    // We're currently under a lock, so tell the continuation to run 
+                    // *asynchronously* so that the TPL does not try to execute it inline
+                    // with this thread.
+                    //
+                    // Note: this flushing is not cancellable.  We've already removed the
+                    // writes from the write queue.  If we were not to write them out we
+                    // would be losing data.
+                    var nextTask = existingWriteTask.ContinueWith(
+                        _ => ProcessWriteQueue(connection, writesToProcess),
+                        CancellationToken.None,
```

Comment:
I'll add a comment, but that would also lead to broken behavior. On shutdown we flush all remaining writes, and then cancel that token. If this talk were still in the thread pool,waiting to be run, then it would be canceled without its writes making it to disk.